ml/synthesis/src/scripts/evaluation.py

```diff
@@ -0,0 +1,353 @@
+MODELS_ROOT = Path(args.models_root)
+model = tf.keras.models.load_model(MODELS_ROOT / mname)
+model.summary()
+
+def new_evaluator(node: NittaNode):
+    final_columns = ['alt_bindings', 'alt_refactorings', 'alt_dataflows', 'pOutputNumber', 'pAlternative', 'pAllowDataFlow', 'pCritical', 'pPercentOfBindedInputs', 'pPossibleDeadlock', 'pNumberOfBindedFunctions', 'pRestless', 'pNotTransferableInputs', 'pRestrictedTime', 'pWaitTime', 'tag_BindDecisionView', 'tag_BreakLoopView', 'tag_ConstantFoldingView', 'tag_DataflowDecisionView', 'tag_OptimizeAccumView', 'tag_ResolveDeadlockView']
+    metrics_columns = [cn for cn in final_columns if cn.startswith("p")] + ["pRefactoringType", "pWave"]
+    
+    node_df = assemble_tree_dataframe("", node, include_label=False, levels_left=-1)
+    filled_metrics_df = pd.concat([pd.DataFrame(columns=metrics_columns), node_df])
+    preprocessed_df = preprocess_df(filled_metrics_df)
+    right_final_columns_df = pd.concat([pd.DataFrame(columns=final_columns), preprocessed_df])[final_columns]
+    ohe_flags_zero_filled_df = right_final_columns_df.fillna(0)
+    final_df = ohe_flags_zero_filled_df
+    
+    return model.predict(final_df.values)[0][0]
+    
+from aiohttp import ServerDisconnectedError
+from collections import defaultdict
+
+def old_evaluator(node: NittaNode):
+    return node.score
+
+counters = defaultdict(lambda: 0)
+def reset_counters():
+    global counters
+    counters = defaultdict(lambda: 0)  
+
+async def select_best_by_evaluator(session, evaluator, node, children_limit=None):
+    counters[evaluator.__name__] += 1
+    
+    if node.is_leaf:
+        if not node.is_finish:
+            return None
+            
+        return node
+
+    try:
+        await node.retrieve_subforest(session, 0)
+    except ServerDisconnectedError:
+#         print(f"Invalid node with NITTA exception: {node}")
+        return None
+    
+    children = [(evaluator(child), child) for child in node.children]
+    children.sort(key=lambda v: v[0], reverse=True)
+#     print(f"children: {[d[0] for d in children]}")
+    if children_limit:
+        children = children[:children_limit]
+    
+    while children:
+        next_best_child = children.pop(0)[1]
+#         print(f"next best: {next_best_child}")
+        result = await select_best_by_evaluator(session, evaluator, next_best_child, children_limit)
+        if result is not None:
+            return result         
+        
+    return None
+    
+async def main():
+    example = args.example
+    
+    reset_counters()
+
+    nitta_tree = None
+    with subprocess.Popen(f"{NITTA_EXE_PATH} -p={NITTA_PORT} {example}", 
+                      cwd=NITTA_ROOT_DIR, stdout=subprocess.PIPE, stderr=subprocess.STDOUT, shell=True
+                     ) as proc:
+        try:
+            time.sleep(WAIT_NITTA_DELAY)
+            print(proc.stdout.read1().decode())
+            nitta_tree = await retrieve_whole_nitta_tree()
+        finally:
+            proc.kill()
```

Comment:
I guess, we could keep the NITTA process (and evaluated tree!) for the second part of the script and avoid starting it the second time. It will let us:
- reuse evaluated nodes (they're cached in RAM once evaluated - will drastically speed this script up)
- save some time by removing second NITTA startup